CRM/Core/BAO/UFGroup.php

```diff
@@ -2014,31 +2014,21 @@ public static function buildProfile(
       $form->add('select', $name, $title, $subtypeList, $required, array('class' => 'crm-select2', 'multiple' => TRUE));
     }
     elseif (in_array($fieldName, CRM_Contact_BAO_Contact::$_greetingTypes)) {
-      //add email greeting, postal greeting, addressee, CRM-4575
-      $gId = $form->get('gid') ? $form->get('gid') : CRM_Utils_Array::value('group_id', $field);
+      // Get contact type for greeting selector
+      $gId = $form->get('gid') ?: CRM_Utils_Array::value('group_id', $field);
       $profileType = CRM_Core_BAO_UFField::getProfileType($gId, TRUE, FALSE, TRUE);
 
-      if (empty($profileType) || in_array($profileType, array(
-        'Contact',
-        'Contribution',
-        'Participant',
-        'Membership',
-      ))
-      ) {
-        $profileType = 'Individual';
+      if (!$profileType || in_array($profileType, ['Contact', 'Contribution', 'Participant', 'Membership'])) {
+        $profileType = ($profileType == 'Contact' && $form->get('id')) ? CRM_Contact_BAO_Contact::getContactType($form->get('id')) : 'Individual';
```

Comment:
OK - I feel like I would probably extract the 'inids' of this function (below the part where $name is determined) and create a sub-function called buildNamedProfile & we could probably restore sanity here.

I did find some reasons to give me enough comfort to merge this though

1) Later in the external_identifier part I see a clear assumption that $form->get('id') would be the contact ID
```
      $form->add('text', $name, $title, $attributes, $required);
      $contID = $contactId;
      if (!$contID) {
        $contID = $form->get('id');
      }
      $form->addRule($name,
        ts('External ID already exists in Database.'),
        'objectExists',
        array('CRM_Contact_DAO_Contact', $contID, 'external_identifier')
      );
    }
```

2) On reflection this parameter is ONLY used for the greeting types so the risk is fairly contained.

3) this needs cleaning up but step one is probably just to reformat the file otherwise it's gonna be a pain